tests/e2e/e2e_test.go

```diff
@@ -193,93 +191,465 @@ func (s *IntegrationTestSuite) TestConcentratedLiquidity() {
+
+	// Assert that the balance changed only for tokenIn (uosmo)
+	s.assertBalancesInvariants(addr1BalancesBefore, addr1BalancesAfter, false, true)
+
+	// Assert Balances: Swap 1
+
+	// Calculate uncollected fees for address1 position1
+	feesUncollectedAddress1Position1_Swap1 := calculateUncollectedFees(
+		positionsAddress1[0].Position.Liquidity,
+		sdk.ZeroDec(), // no growth below
+		sdk.ZeroDec(), // no growth above
+		sdk.ZeroDec(), // no feeGrowthInsideLast - it is the first swap
+		feeGrowthGlobal,
+	)
+
+	// Assert
+	s.Require().Equal(
+		addr1BalancesBefore.AmountOf("uosmo").Add(feesUncollectedAddress1Position1_Swap1.TruncateInt()),
+		addr1BalancesAfter.AmountOf("uosmo"),
+	)
+
+	// Swap 2
+	//
+	// Cross-tick swap:
+	// * Part of swap happens in range of liquidity of 3 positions: both of address1 and one for address3 (until tick 400 - upper tick of address1 position1)
+	// * Another part happens in range of liquidity of 2 positions: one from address1 and address3
+	//
+	// Asserts:
+	// * Net liquidity is kicked out when crossing initialized tick
+	// * Liquidity of position that was kicked out after first swap step does not earn rewards from second swap step
+	// * Uncollected fees from multiple swaps are correctly summed up and collected
+
+	// tickOffset is a tick index after the next initialized tick to which this swap needs to move the current price
+	tickOffset := sdk.NewInt(3)
+	sqrtPriceBeforeSwap = concentratedPool.GetCurrentSqrtPrice()
+	liquidityBeforeSwap = concentratedPool.GetLiquidity()
+	nextInitTick := sdk.NewInt(400) // address1 position1's upper tick
+
+	// Calculate sqrtPrice after and at the next initialized tick (upperTick of address1 position1 - 400)
+	sqrtPriceAfterNextInitializedTick, err := cl.TickToSqrtPrice(nextInitTick.Add(tickOffset), sdk.NewInt(precisionFactorAtPriceOne))
+	s.Require().NoError(err)
+	sqrtPriceAtNextInitializedTick, err := cl.TickToSqrtPrice(nextInitTick, sdk.NewInt(precisionFactorAtPriceOne))
+	s.Require().NoError(err)
+
+	// Calculate Δ(sqrtPrice):
+	// deltaSqrtPriceAfterNextInitializedTick = ΔsqrtP(403) - ΔsqrtP(400)
+	// deltaSqrtPriceAtNextInitializedTick = ΔsqrtP(400) - ΔsqrtP(currentTick)
+	deltaSqrtPriceAfterNextInitializedTick := sqrtPriceAfterNextInitializedTick.Sub(sqrtPriceAtNextInitializedTick)
+	deltaSqrtPriceAtNextInitializedTick := sqrtPriceAtNextInitializedTick.Sub(sqrtPriceBeforeSwap)
+
+	// Calculate the amount of osmo required to:
+	// * amountInToGetToTickAfterInitialized - move price from next initialized tick (400) to destination tick (400 + tickOffset)
+	// * amountInToGetToNextInitTick - move price from current tick to next initialized tick
+	// Formula is as follows:
+	// Δy = L * Δ(sqrtPrice)
+	amountInToGetToTickAfterInitialized := deltaSqrtPriceAfterNextInitializedTick.Mul(liquidityBeforeSwap.Sub(positionsAddress1[0].Position.Liquidity))
+	amountInToGetToNextInitTick := deltaSqrtPriceAtNextInitializedTick.Mul(liquidityBeforeSwap)
```

Comment:
just wondering if we can put these calculation in helper function, since i've seen it elsewhere in code too. And it'll make it cleaner too :))

Reply:
hmm, it is possible to put it into a helper function, but these calculations are calculated only once, so we might do that only for readability purposes, wdyt?

PS: swap 2 and swap 3 have similar calculations, but not same, so not possible to use the same function for both